doc/private/man/man3/CliBase.3
.TH "CliBase" 3 "Wed Feb 8 2012" "Version 0.0.0" "EMI File Transfer Service (FTS 3)" \" -*- nroff -*-
.ad l
.nh
.SH NAME
CliBase \- 
.SH SYNOPSIS
.br
.PP
.PP
Inherited by \fBSubmitTransferCli\fP.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "void \fBinitCli\fP (int ac, char *av[])"
.br
.ti -1c
.RI "bool \fBprintHelp\fP ()"
.br
.ti -1c
.RI "bool \fBprintVersion\fP ()"
.br
.ti -1c
.RI "bool \fBisVerbose\fP ()"
.br
.ti -1c
.RI "bool \fBisQuite\fP ()"
.br
.ti -1c
.RI "string \fBgetService\fP ()"
.br
.ti -1c
.RI "string \fBgetSource\fP ()"
.br
.ti -1c
.RI "string \fBgetDestination\fP ()"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "string \fBdiscoverService\fP ()"
.br
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "variables_map \fBvm\fP"
.br
.ti -1c
.RI "options_description \fBbasic\fP"
.br
.ti -1c
.RI "options_description \fBcli_options\fP"
.br
.ti -1c
.RI "positional_options_description \fBp\fP"
.br
.ti -1c
.RI "string \fBendpoint\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 16 of file CliBase.h.

.SH "Author"
.PP 
Generated automatically by Doxygen for EMI File Transfer Service (FTS 3) from the source code.


fts2-prototype/RELEASE-NOTES
*** Release glite-data-transfer-cli_R_4_0_2 2011-10-14
	#109 	glite-transfer-submit man page issues https://svnweb.cern.ch/trac/glitefts/ticket/109

*** Release glite-data-transfer-cli_R_4_0_0 2011-03-07

    o glite-transfer-channel-audit fails with Segmentation Fault  (bug#68524)
    o Tool to modified automatically the state of the channels given an SE  (bug#61780)    


*** Release glite-data-transfer-cli_R_3_7_5_1 2010-11-29

    o changed util.c to output from stdout to stderr  (bug#75722)

*** Release glite-data-transfer-cli_R_3_7_4_1 2010-01-22

    o added glite-transfer-se-set (bug#61780)

*** Release glite-data-transfer-cli_R_3_7_3_1 2009-12-17

	o Fix for bug #52416: Bad error treatment for ExistsException on the client	side
	o Minimal Python API with test.

*** Release glite-data-transfer-cli_R_3_7_2_1 2009-08-28

    o bug #41584: [delegation] IPv6 bug: non compliant name resolving 
      function in source code (gethostbyname_r)
      This bug also applies to this component. 

*** Release glite-data-transfer-cli_R_3_7_1_1 2009-08-05

	o Fixing a regression: overwrite flag (-o) should not require an argument
    o Updated the test suite to the latest FTS service.

*** Release glite-data-transfer-cli_R_3_7_0_1 2009-07-03

	o v3.7.0 of the interface with checksum support.

	  glite-transfer-submit can take the -K option for checksum comparison
	  and can pass on optional checksums (algorithm:hexadecimal-numbers)
	  for each source-destination SURL pair given by the command line or
	  in a job description file.

*** Release glite-data-transfer-cli_R_3_6_3_1 2009-03-30
	o fixed bug #48684: transfer-cli should obsolete transter-api-c and srm-cli packages

*** Release glite-data-transfer-cli_R_3_6_2_1 2009-03-22
    
	o Fixed Savannah bug #35997:  FTS cli: if an invalid value is passed
	  with -s option, the default is used
    o Removed obsolete commands: glite-transfer-placement-submit
      and glite-transfer-discovery

*** Release glite-data-transfer-cli_R_3_6_0_1 2009-01-09
    o fixed bug 45167
    o fixed bug 28212
    o fixed bug #39819
	o added site group management tools
	o Fixed the seg fault for --tx-log-level and --srmcopy-direction options
	o Fixed the option --urlcopy-tx-to-per-mb that was executing --urlcopy-no-prog-to
	o Fixed libraries build order (avoid relinking and appending U to .so file names)

*** Release glite-data-transfer-cli_R_3_5_2_1 2008-12-04
	o Fixed bug 41682

*** Release glite-data-transfer-cli_R_3_5_1_1 2008-12-03
	o slc5 fixes

*** Release glite-data-transfer-cli_R_3_5_0_1 2008-11-27
    o Merged transfer-api-c into transfer-cli.
    o updated to schema version 3.5.0

*** Release glite-data-transfer-cli_R_3_4_5_1 2008-08-11
    o fixed bug #40064 
	o fixed bug #39538
    o refactored options parsing 
    o modified help messages

*** Release glite-data-transfer-cli_R_3_4_4_3 2008-06-17
    o Added dependency from gsoap

*** Release glite-data-transfer-cli_R_3_4_4_2 2008-05-08
	o Fix bug #33406

*** Release glite-data-transfer-cli_R_3_4_3_2 2008-03-20
	o changes to command line help in submit - clearer token handling

*** Release glite-data-transfer-cli_R_3_4_3_1 2008-03-18
	 #30034: FTS: cannot set numerical channel properties to zero
	 #33406: FTS: CLI version string is not reported correctly in verbose option
	 #28524: Missing -o parameter from glite-transfer-submit help
	 Interface changes related to the latest schema changes 3.2.0
	 Number of undocumented bugs fixed related to the marshalling.

*** Release glite-data-transfer-cli_R_3_4_2_2 2007-06-06
    o lib64 changes

*** Release glite-data-transfer-cli_R_3_4_2 2007-06-01

   o Added overwrite flag for SRM 2.2. Proper name for space token description.

*** Release glite-data-transfer-cli_R_3_4_1-1 2007-04-20

   o Fixed SLC4 build error and a few compiler warnings.

*** Release glite-data-transfer-cli_R_3_3_0-4 2007-03-02
   o Compile with gsoap 2.7.6b and 2.7.9b 

*** Release glite-data-transfer-cli_R_3_3_0-1 2006-08-22
   o Add option -x to the glite-transfer-discovery tool in order to display all the 
     FTS service information
   o Add option --all to the glite-transfer-discovery tool in order to return
     all the FTS services that are able to manage a transfer between the given two sites
   o Fix for #19196: FTS: cli should support site names in teh -s option

*** Release glite-data-transfer-cli_R_3_2_3_4 2006-07-06

   o Fix for #17975 - service info interface version bad
   o Fix for #16546 - bad error on addChannel
   o Make interesting states (default serach) only Active, Submitted and Pending
   o Better date format on stats
   o Fix for #16581 - unclear manpage for addManager

*** Release glite-data-transfer-cli_R_3_2_3-3 2006-06-21

	o Merge changes from 3.2.2-3

*** Release glite-data-transfer-cli_R_3_2_2-3 2006-06-21
    Fixed Bugs:
     o #17608: FTS: glite-transfer-stats-vo segfaults

*** Release glite-data-transfer-cli_R_3_2_3-2 2006-06-14

	o Merge changes from 3.2.2-2

*** Release glite-data-transfer-cli_R_3_2_2-2 2006-06-14
    Fixed Bugs:
     o #17460: FTS: glite-transfer-discovery display the wrong destination site when the --verbose option is used

*** Release glite-data-transfer-cli_R_3_2_3 2006-05-02

	o Added Doxygen API documentation for the simple transfer API.

*** Release glite-data-transfer-cli_R_3_2_2 2006-04-28
    Add command line options -T and -S to glite-transfer-discovery
    Fixed Bugs:
     o #16424 FTS: BDII service discovery should work 
     o #16215: fix man page of glite-transfer-channel-list 

*** Release glite-data-transfer-cli_R_3_2_1 2006-03-23

    o CGSI_gSOAP library location has changed from ${repository}/${ext.cgsi-gsoap.subdir} to ${with.cgsi-gsoap.prefix}

*** Release glite-data-transfer-cli_R_3_2_0 2006-02-22
    Add glite-transfer-discovery tool for returning the endpoint of an
       FTS service that manage a channel between two sites
    Fixed bugs:
     o #14765 FTS: Logical name is not displayed for placement jobs


src/CLI/CliBase.cpp
/*
 * CliBase.cpp
 *
 *  Created on: Feb 2, 2012
 *      Author: simonm
 */

#include "CliBase.h"
#include  "CliManager.h"

#include <iostream>

#include "ServiceDiscoveryIfce.h" // check "" or <>
#include <glite/data/glite-util.h>
#include <boost/lexical_cast.hpp>
#include <boost/scoped_array.hpp>

using namespace boost;

CliBase::CliBase() : basic("Basic options") {
    basic.add_options()
			("help,h", "Print this help text and exit.")
			("quite,q", "Quiet operation.")
			("verbose,v", "Be more verbose.")
			("service,s", value<string>(), "Use the transfer service at the specified URL.")
			("source", value<string>(), "Specify source site name.")
			("destination", value<string>(), "Specify destination site name.")
			("version,V", "Print the version number and exit.")
			;

    cli_options.add(basic);

	p.add("source", 1);
	p.add("destination", 1);
}

CliBase::~CliBase() {

}


void CliBase::initCli(int ac, char* av[]) {
	// turn off guessing, so --source is not mistaken with --source-token
	int style = command_line_style::default_style & ~command_line_style::allow_guessing;

	store(command_line_parser(ac, av).options(cli_options).positional(p).style(style).run(), vm);
	notify(vm);

	if (vm.count("service")) {
		endpoint = vm["service"].as<string>();
		// check if the endpoint has the right prefix
		if (endpoint.find("http") != 0 && endpoint.find("https") != 0 && endpoint.find("httpd") != 0) {
			// if not erase
			cout << "Wrongly formatted endpoint: " << endpoint << endl;
			endpoint.erase();
		}
	} else {
		endpoint = discoverService();
	}
}

bool CliBase::printHelp() {

	if (vm.count("help")) {
        cout << cli_options << endl;
        return true;
    }

    return false;
}

bool CliBase::printVersion() {

	if (vm.count("version")) {
        cout << "TODO" << endl;
        return true;
    }

    return false;
}

bool CliBase::isVerbose() {
	return vm.count("verbose");
}

bool CliBase::isQuite() {
	return vm.count("quite");
}

string CliBase::getService() {

	return endpoint;
}

string CliBase::getSource() {

	if (vm.count("source")) {
		return vm["source"].as<string>();
	}
	return "";
}

string CliBase::getDestination() {

	if (vm.count("destination")) {
		return vm["destination"].as<string>();
	}
	return "";
}

string CliBase::discoverService() {

	CliManager* manager = CliManager::getInstance();

	string source = getSource(), destination = getDestination();
	SDService *service;
	scoped_array<char> c_str_arr, error_arr;
	char *error, *c_str = 0;
	const char * fts_srvtype;
	const char * fts_version;
	SDException exc;

	if (source.size() > 0 && destination.size() > 0) {

		SDServiceData datas[] =
		{
			{"source", (char*) source.c_str()},
			{"destination", (char*) destination.c_str()}
		};

		SDServiceDataList datalist = {NULL, 2, datas};

		c_str = glite_discover_service_by_data(manager->FTS3_CA_SD_TYPE.c_str(), &datalist, &error);
		c_str_arr.reset(c_str);
		error_arr.reset(error);
	}

	if (getenv(manager->FTS3_SD_ENV.c_str()))
		fts_srvtype = getenv(manager->FTS3_SD_ENV.c_str());
	else
		fts_srvtype = manager->FTS3_SD_TYPE.c_str();

	if(getenv(manager->FTS3_IFC_VERSION.c_str()))
		fts_version = getenv(manager->FTS3_IFC_VERSION.c_str());
	else
		fts_version = manager->FTS3_INTERFACE_VERSION.c_str();

	// Note: glite_discover_service will check if the service name passed
	// has an associated service of the given type (fts_srvtype), if not
	// it will check if the name refers to a site name or an host name

	c_str = glite_discover_service_by_version(fts_srvtype, c_str, fts_version, &error);
	c_str_arr.reset(c_str);
	error_arr.reset(error);

	if (!c_str) {
        cout << "Service discovery: " << error << endl;
		return "";
	}

	service = SD_getService(c_str, &exc);
	if (!service) {
		cout << "Service discovery lookup failed for " << c_str << ": " << exc.reason << endl;
		SD_freeException(&exc);
		return "";
	}

	string tmp;
	if (service->endpoint) {
		tmp = service->endpoint;
	}
	SD_freeService(service);

	return tmp;
}
